### PR TITLE
add FieldKey type

### PR DESCRIPTION
**Pull request by tanner0101**
Adds a new top-level `FieldKey` type to FluentKit. This type represents possible field key values but in a slightly more type-safe way than `String` is currently able to. 

⚠️ This change removes the ability to use custom ID keys on models. `@ID`'s key parameter must be  either `"id"` (for backward compatibility) or the new `.id`. Additional identifying fields can be added with custom keys using `@Field` with a unique constraint. Read on for more information on this decision.

This is what the _basic_ structure of `FieldKey` looks like:

```swift
public enum FieldKey: ExpressibleByStringLiteral {
    case id
    case string(String)
}
```

The important piece here, and the reason for this change, is that `FieldKey.id` is a special type of key that does not have an actual string value. This key can be interpreted by the underlying driver and converted to the correct value. For Mongo this would be `_id` and for SQL it would be `id`. This is critical for defining models that can work across all of Fluent's supported driver types.

`FieldKey` is usable everywhere String keys were previously usable, including property wrappers, filter expressions, and schema builders.

```swift
final class Planet: Model {
    static let schema = "planets"

    @ID(key: .id)
    var id: UUID?
    
    @Field(key: "name")
    var name: String
}

Planet.query(on: database).filter(.id, .equals, ...)

struct CreatePlanet: Migration {
    func prepare(on database: Database) -> EventLoopFuture<Void> {
        database.schema("planets")
            .field(.id, .uuid, .identifier(auto: false)
            .field("name", .string, .required)
            .create()
    }

    func revert(on: database)  -> EventLoopFuture<Void> {
        database.schema("planets").delete()
    }
}
```

`FieldKey` neatly solves the problem of dynamic ID keys and it may also solve future problems where there are differences in keys between databases. For example, the currently hacky `"fluentAggregate"` key could be a part of this enum.

One other potentially useful aspect of `FieldKey`  is that it can be extended by the end user. This could be a good solution to keeping String key declarations DRY. For example:

```swift
extension FieldKey {
    static var name: Self { "name" }
}
```

Assuming this extension, we could refactor the previous `Planet` code to look like this:

```swift
final class Planet: Model {
    static let schema = "planets"

    @ID(key: .id)
    var id: UUID?
    
    @Field(key: .name)
    var name: String
}

Planet.query(on: database).filter(.name, .equals, "Earth")

struct CreatePlanet: Migration {
    func prepare(on database: Database) -> EventLoopFuture<Void> {
        database.schema("planets")
            .field(.id, .uuid, .identifier(auto: false)
            .field(.name, .string, .required)
            .create()
    }

    func revert(on: database)  -> EventLoopFuture<Void> {
        database.schema("planets").delete()
    }
}
```

### CODE REVIEW

**Comment by Joannis** [comment]
Love it

File: Sources/FluentKit/Migration/MigrationLog.swift [hunk 1]
```diff
@@ -8,7 +8,7 @@ public final class MigrationLog: Model {
         return MigrationLogMigration()
     }
 
-    @ID(key: "id")
+    @ID(key: .id)
     public var id: Int?
```
Comment (Joannis) [inline Sources/FluentKit/Migration/MigrationLog.swift]: This should be using `UUID` for MongoDB support

**Comment by 0xTim** [comment]
@tanner0101 love it! I've been using a similar thing (enums backed by `String`s to put all the names in one place) in projects, but having it baked into Fluent allows it to be supported more naturally.

The only thing I'd suggest is for the example, not calling the property `name` as that confused me for a while! I'm assuming the `FieldKey` has a case for `name` since that's what the property wrapper expects, not because that's what the property in the model class is 😅 

Final thing to note is we should be careful about using an enum for something we may want to extend in the future. I can't remember if the frozen enum stuff allows us to mark the enum are non-frozen (I think that was for the standard library only) but we don't want to end up in a situation where a new Fluent property wrapper comes along and can't be used until Fluent 5 because we have an enum

**Comment by tanner0101** [comment]
@0xTim good points. `name` is definitely confusing, you're right. Maybe `string` would be better. Regarding the enum stuff, no unfortunately the library resilience stuff is not available to SPM package since they compile source files. I'm not sure what the best solution is here. 

The safest option is to do something like:
```swift
public struct FieldKey {
    public static var id: Self { 
        .init(storage: .id) 
    }

    public static func string(_ string: String): Self { 
        .init(storage: .string(string) 
    }   

    public var isID: Bool { 
        guard case .id = self.storage else { return false }
        return true
    }

    public var asString: String? { 
        guard case .string(let string) = self.storage else { return nil }
        return string
    }
    internal enum Storage {
        case id
        case string(String)
    }
    var storage: Storage
}
```

I'm really not a fan of this though because it makes the code hard to read and prevents anything non-internal from switching over the enum (Internal things can check `storage`). 

I'd really like to avoid going back down that route like some Vapor 3 packages did if at all possible. Maybe we could adopt a convention for letting the user know which enums are frozen and safe to switch over. There's not really a reason why an end user should need to switch over an enum like this, so if there was some way of letting them know switching is "do at your own risk" and we are free to add new cases that could be a good compromise.

**Comment by tanner0101** [comment]
Merging this into the beta-5 branch where we can continue to discuss it alongside other updates.

**Pull request merged**